src/plugins/mer/meremulatormodeoptionswidget.cpp

```diff
@@ -0,0 +1,233 @@
+{
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.dconf() != value) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 model.displayResolution(),
+                                                 model.displaySize(),
+                                                 value,
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelScreenResolutionChanged(const QSize &resolution)
+{
+    const auto name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.displayResolution() != resolution) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 resolution,
+                                                 model.displaySize(),
+                                                 model.dconf(),
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::deviceModelScreenSizeChanged(const QSize &size)
+{
+    const auto name = m_ui->deviceModelComboBox->currentDeviceModel();
+    const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+    QTC_ASSERT(!model.isNull(), return);
+
+    if (model.displaySize() != size) {
+        updateDeviceModel(MerEmulatorDeviceModel(name,
+                                                 model.displayResolution(),
+                                                 size,
+                                                 model.dconf(),
+                                                 model.isSdkProvided()));
+    }
+}
+
+void MerEmulatorModeOptionsWidget::updateGui()
+{
+    if (m_ui->deviceModelComboBox->count() == 0)
+        return;
+
+    const QString name = m_ui->deviceModelComboBox->currentDeviceModel();
+    if (m_deviceModels.keys().contains(name)) {
+        const MerEmulatorDeviceModel model = m_deviceModels.value(name);
+        m_ui->removeProfileButton->setDisabled(model.isSdkProvided());
+        m_ui->optionsWidget->setCurrentDeviceModel(model);
+    }
+
+}
+
+void MerEmulatorModeOptionsWidget::on_addProfileButton_clicked()
+{
+    QTC_CHECK(m_deviceModels.uniqueKeys()
+              == m_deviceModels.keys());
+
+    // Find number for new device model
+    const QSet<QString> names = m_deviceModels.keys().toSet();
+    const QString name = MerEmulatorDeviceModel::uniqueName(tr("Unnamed"), names);
+
+    if (!addDeviceModel(name)) {
+        QMessageBox::warning(this, tr("Create device model warning"), tr("Emulator device model is not created.\n"
+                                                                    "Provided name \"%1\" is not unique.").arg(name));
```

Comment:
Given the `uniqueName()` use this shouldn't be among the possible reasons of failure now, right?

Reply:
Right. Message removed.